src/air/constraints/evaluator.rs

```diff
@@ -144,12 +145,36 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
             let mut transition_adjustment = Vec::with_capacity(lde.len());
             for d in lde.iter() {
                 let degree_adjustment = self.air.composition_poly_degree_bound()
-                    - (self.air.context().trace_length * (transition_degree - 1));
+                    - (trace_length * (transition_degree - 1));
                 transition_adjustment.push(d.pow(degree_adjustment));
             }
             degree_adjustments.push(transition_adjustment);
         }
 
+        let x_n = Polynomial::new_monomial(FieldElement::<F>::one(), trace_length);
+        let x_n_1 = x_n - FieldElement::<F>::one();
+
+        let mut zerofier_evaluations = evaluate_polynomial_on_lde_domain(
+            &x_n_1,
+            domain.blowup_factor,
+            domain.interpolation_domain_size,
+            &domain.coset_offset,
+        )
+        .unwrap();
+
+        FieldElement::inplace_batch_inverse(&mut zerofier_evaluations);
+        let transition_zerofiers_inverse_evaluations: Vec<Vec<FieldElement<F>>> =
+            transition_exemptions_evaluations
+                .iter()
+                .map(|row| {
+                    zerofier_evaluations
+                        .iter()
+                        .zip(row.iter())
+                        .map(|(c1, c2)| c1.clone() * c2.clone())
```

Comment:
I think those clones are unnecessary.

Reply:
we need that clones because multiplication needs it and we have references at that point

Reply:
sorry, I've just changed that and it worked